effectful-th/src/Effectful/TH.hs

```diff
@@ -0,0 +1,243 @@
+  --   variable and substitute it later for 'Eff es'.
+  --
+  -- - A variable 'es' for the local 'Eff es' if the monad parameter was locally
+  --   substituted in the contructor.
+  --
+  -- For example in the following effect:
+  --
+  -- data E :: Effect where
+  --   E1 :: Int -> E m ()
+  --   E2 :: IOE :> es => E (Eff es) ()
+  --
+  -- Processing 'E1' will yield 'Right m', but 'E2' will yield 'Left es'.
+  --
+  -- In the first case we need to substitute the variable ourselves in a few
+  -- places, but in the second we're good since it was already substituted.
+  (esName, maybeMonadName) <- case ename of
+    Left  esName    -> pure (esName, Nothing)
+    Right monadName -> (, Just monadName) <$> newName "es"
+
+  let fnName = mkName . toSmartConName $ nameBase name
+  fnArgs <- traverse (const $ newName "x") actionParams
+
+  let esVar = VarT esName
+
+      substM :: Type -> Type
+      substM = case maybeMonadName of
+        Just m  -> applySubstitution . Map.singleton m $ ConT ''Eff `AppT` esVar
+        Nothing -> id
+
+      (origActionVars, actionCtx) = extractCtx typ
+      actionVars = case maybeMonadName of
+        Just m  -> filter ((m /=) . tvName) origActionVars
+                ++ [kindedTVSpecified esName $ ListT `AppT` ConT ''Effect]
+        Nothing -> origActionVars
+
+#if MIN_VERSION_template_haskell(2,17,0)
+  -- In GHC >= 9.0 it's possible to generate the following body:
+  --
+  -- e x1 .. xN = send (E @ty1 .. @tyN x1 .. xN)
+  --
+  -- because specificities of constructor variables are exposed.
+  --
+  -- This allows to generate functions for such effects:
+  --
+  -- type family F ty :: Type
+  -- data AmbEff :: Effect where
+  --   AmbEff :: Int -> AmbEff m (F ty)
+  --
+  -- Sadly the version for GHC < 9 will not compile due to ambiguity error.
+  let fnBody =
+        let tvFlag = \case
+              PlainTV  _ flag   -> flag
+              KindedTV _ flag _ -> flag
+
+            tyApps = (`mapMaybe` origActionVars) $ \v -> case tvFlag v of
+              InferredSpec  -> Nothing
+              SpecifiedSpec -> Just $ if maybeMonadName == Just (tvName v)
+                                      then ConT ''Eff `AppT` esVar
+                                      else VarT (tvName v)
+
+            effCon = if makeSig
+              then foldl' AppTypeE (ConE name) tyApps
+              else                  ConE name
+        in VarE 'send `AppE` foldl' (\f -> AppE f . VarE) effCon fnArgs
+#else
+  -- In GHC < 9.0, generate the following body:
+  --
+  -- e :: E v1 .. vN :> es => x1 -> .. -> xK -> E v1 .. vN (Eff es) r
+  -- e x1 .. xK = send (E x1 .. xN :: E v1 .. vK (Eff es) r)
+  let fnBody =
+        let effOp  = foldl' (\f -> AppE f . VarE) (ConE name) fnArgs
+            effSig = effTy `AppT` (ConT ''Eff `AppT` esVar) `AppT` substM resTy
+        in if makeSig
+           then VarE 'send `AppE` SigE effOp effSig
+           else VarE 'send `AppE`      effOp
+#endif
+  let fnSig = ForallT actionVars
```

Comment:
If we use explicit quantification, datatype heads without kind annotations for their parameters like `data Effect m a` will fail with confusing type ambiguity errors under `PolyKinds`. One way to solve this is to scrap explicit tyvars and let the inference do the job, at the cost of not being able to generate some ambiguous constructors. What do you think?

Reply:
They will fail, but there is a check earlier in `makeEffectImpl` that validates kinds of the last two type parameters, so if you skip it, you will get a reasonable error message:

```
tests/ThTests.hs:21:1: error:
    Expected the next to last type parameter to have a kind * -> *, got k_0
   |
21 | makeEffect ''SimpleADT
```

```
tests/ThTests.hs:21:1: error:
    Expected the last type parameter to have a kind *, got k_0
   |
21 | makeEffect ''SimpleADT
   | ^^^^^^^^^^^^^^^^^^^^^^
```

This encourages writing effects like this:

```
data E x1 ... xN :: Effect where
   ...
```

which I find more informative :slightly_smiling_face: